### PR TITLE
Defer Attestations for Future Processing in Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR adds the ability to defer processing of attestations that will be valid in the future in our optimized slasher implementation. The way it works is:

1. Every epoch, get a set of attestations to process from the queue
2. Filter out the attestations which are valid, invalid, and valid in the future
3. Put the ones valid in the future back into the queue
4. Throw away the invalid ones
5. Process the valid ones

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/slasher/receive_test.go [hunk 1]
```diff
@@ -331,40 +329,7 @@ func TestSlasher_receiveAttestations_OK(t *testing.T) {
 		att1,
 		att2,
 	}
-	require.DeepEqual(t, wanted, s.attestationQueue)
-}
-
-func TestSlasher_receiveAttestations_OnlyValidAttestations(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/slasher/receive_test.go:337]: Test no longer relevant as validation no longer happens in receiveAttestations but rather in processQueuedAttestations

File: beacon-chain/slasher/queue.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
+package slasher
```
Comment (rauljordan) [inline beacon-chain/slasher/queue.go:1]: Simple layer of abstraction which makes the code cleaner in other places

File: beacon-chain/slasher/receive.go [hunk 3]
```diff
@@ -69,27 +61,33 @@ func (s *Service) processQueuedAttestations(ctx context.Context, epochTicker <-c
...
 		select {
 		case currentEpoch := <-epochTicker:
-			s.attestationQueueLock.Lock()
-			attestations := s.attestationQueue
-			s.attestationQueue = make([]*slashertypes.IndexedAttestationWrapper, 0)
-			s.attestationQueueLock.Unlock()
+			// We take all the attestations in the queue and filter out
+			// those which are valid now and valid in the future.
+			attestations := s.attsQueue.dequeue()
+			validAtts, validInFuture, numDropped := s.validateAttestationIntegrity(attestations, currentEpoch)
```
Comment (0xKiwi) [inline beacon-chain/slasher/receive.go:67]: Would this mean we process the same future attestations validity multiple times?

File: beacon-chain/slasher/helpers.go [hunk 4]
```diff
@@ -47,6 +46,55 @@ func (s *Service) groupByChunkIndex(
...
 }
 
+// Validates the attestation data integrity, ensuring we have no nil values for
+// source, epoch, and that the source epoch of the attestation must be less than
+// the target epoch, which is a precondition for performing slashing detection.
+// This function also checks the attestation source epoch is within the history size
+// we keep track of for slashing detection.
+// This function returns a list of valid attestations, a list of attestations that are
+// valid in the future, and the number of attestations dropped.
+func (s *Service) validateAttestationIntegrity(
```
Comment (terencechain) [inline beacon-chain/slasher/helpers.go:56]: Im not a fan of doing filtering and returning processed items in a func that beings with `validate` 

A better way would be filter this into `validateAttestationIntegrity` and `filterAttestation`

File: beacon-chain/slasher/queue.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+import (
+	"sync"
+
+	slashertypes "github.com/prysmaticlabs/prysm/beacon-chain/slasher/types"
+)
+
+// Struct for handling a thread-safe list of indexed attestation wrappers.
+type attestationsQueue struct {
+	lock  sync.RWMutex
```
Comment (terencechain) [inline beacon-chain/slasher/queue.go:11]: You can just do `sync.RWMutex`

Then you can just do `q.Lock()` or `q.UnLock()`

**Review by 0xKiwi — APPROVED** [review]

**Pull request merged**